docs/11_templating.md

```diff
@@ -0,0 +1,646 @@
+
+### Custom Processing with Hooks
+
+For advanced file transformations beyond Handlebars templating, use `preHook` and `postHook` scripts. These hooks have full access to the output directory and can perform any custom processing needed.
+
+**Example:**
+```json
+{
+  "generation": {
+    "preHook": "scripts/prepare.js",
+    "postHook": "scripts/finalize.js",
+    "fileTransforms": [
+      { "pattern": "**/*.swift", "processor": "handlebars" }
+    ]
+  }
+}
+```
+
+The `postHook` can process any generated files, including those that were copied without processing. Hooks receive context about the generation process including paths and variables. See the Generation Pipeline section below for more details on hook execution order.
+
+---
+
+## API Usage
+
+### Programmatic Interface
+
+```typescript
+import { 
+  TemplateRegistry, 
+  TemplateGenerator 
+} from '@salesforce/magen-templates';
+
+// 1. Discover available templates
+const registry = new TemplateRegistry();
+const allTemplates = await registry.discoverTemplates();
+
+// 2. Filter by platform (required)
+const iosTemplates = allTemplates.filter(t => t.platform.type === 'ios');
+
+// 3. Filter by capabilities (optional)
+const matchingTemplates = iosTemplates.filter(t => 
+  t.capabilities.includes('offline-sync')
+);
+
+// 4. Examine template metadata to make informed selection
+for (const template of matchingTemplates) {
+  console.log(`Template: ${template.displayName}`);
+  console.log(`Description: ${template.description}`);
+  console.log(`Use Case: ${template.useCase.primary}`);
+  console.log(`Capabilities: ${template.capabilities.join(', ')}`);
+  
+  // Get full metadata including extension points and variables
+  const metadata = await registry.getMetadata(template.id);
+  console.log(`Extension Points: ${metadata.extensionPoints?.length || 0}`);
+  console.log(`Required Variables:`, 
+    metadata.templateVariables.filter(v => v.required).map(v => v.name)
+  );
+}
+
+// 5. AI agent or developer selects appropriate template
+const selectedTemplateId = 'ios-native-swift'; // Based on analysis above
+
+// 6. Get full metadata for generation
+const metadata = await registry.getMetadata(selectedTemplateId);
+
+// 7. Generate project
+const generator = new TemplateGenerator(registry);
+const result = await generator.generate({
+  templateId: selectedTemplateId,
+  metadata,
+  variables: {
+    projectName: 'MyApp',
+    organization: 'My Company',
+    bundleIdentifier: 'com.example.myapp'
+  },
+  outputPath: './MyApp'
+});
+```
+
+### CLI Interface
+
+```bash
+# List templates
+npx magen-templates list
+
+# Show template details
+npx magen-templates info ios-native-swift
+
+# Generate project
+npx magen-templates generate \
+  --template ios-native-swift \
+  --output ./MyApp \
+  --projectName MyApp \
+  --organization "My Company"
+
+# Validate template
+npx magen-templates validate ios-native-swift
```

Comment:
What is the `validate` command validating?

Reply:
this validates the template metadata vs the metadata schema, maybe not something that needs to be exposed through the cli but I have a commit task that runs anytime either (1) the schema changes or (2) metadata is added/modified to ensure that the metadata is compliant with the schema